File: java/lab5/src/productcreation/PersonBuilder.java
package src.productcreation;

import src.baseabstractions.Builder;
import src.baseobjects.Color;
import src.baseobjects.Country;
import src.baseobjects.Person;

/**
 * Builder class for constructing Person objects with a fluent interface.
 *
 * @param <T> the type of Person object to be built, must extend Person
 */
public class PersonBuilder<T extends Person> implements Builder<T> {
    private T person;

    /**
     * Constructs a PersonBuilder initialized with the specified person.
     *
     * @param person the person object to be used as the base for building
     */
    public PersonBuilder(T person) {
        this.person = person;
    }

    /**
     * Sets the name of the person.
     *
     * @param name the name to set
     * @return this builder instance for method chaining
     */
    public PersonBuilder<T> setName(String name) {
        person.setName(name);
        return this;
    }

    /**
     * Sets the height of the person.
     *
     * @param height the height to set
     * @return this builder instance for method chaining
     */
    public PersonBuilder<T> setHeight(long height) {
        person.setHeight(height);
        return this;
    }

    /**
     * Sets the hair color of the person.
     *
     * @param hairColor the hair color to set
     * @return this builder instance for method chaining
     */
    public PersonBuilder<T> setHairColor(Color hairColor) {
        person.setHairColor(hairColor);
        return this;
    }

    /**
     * Sets the nationality of the person.
     *
     * @param nationality the nationality to set
     * @return this builder instance for method chaining
     */
    public PersonBuilder<T> setNationality(Country nationality) {
        person.setNationality(nationality);
        return this;
    }

    /**
     * Resets the builder to its initial state.
     * Currently does nothing as the builder uses an existing person reference.
     */
    public void reset() {

    }

    /**
     * Returns the constructed person object.
     *
     * @return the built person of type T
     */
    public T create() {
        return person;
    }
}

File: java/lab8/client/src/main/kotlin/objectCreation/PersonBuilder.kt
package objectCreation

import abstractions.Builder
import objects.Color
import objects.Country
import objects.Person

/**
 * Builds a [Person] instance field by field.
 */
class PersonBuilder : Builder<Person> {

    private var name: String = ""
    private var height: Long = 1
    private var hairColor: Color = Color.GREEN
    private var nationality: Country? = null

    override fun reset() {
        name = ""
        height = 1
        hairColor = Color.GREEN
        nationality = null
    }

    /**
     * Sets the person's [name].
     *
     * @param name non-blank name
     * @return this builder
     */
    fun setName(name: String): PersonBuilder {
        this.name = name
        return this
    }

    /**
     * Sets the person's [height] in centimetres.
     *
     * @param height must be greater than 0
     * @return this builder
     */
    fun setHeight(height: Long): PersonBuilder {
        this.height = height
        return this
    }

    /**
     * Sets the person's [hairColor].
     *
     * @param hairColor the hair color value
     * @return this builder
     */
    fun setHairColor(hairColor: Color): PersonBuilder {
        this.hairColor = hairColor
        return this
    }

    /**
     * Sets the person's [nationality], which may be null.
     *
     * @param nationality the country or null
     * @return this builder
     */
    fun setNationality(nationality: Country?): PersonBuilder {
        this.nationality = nationality
        return this
    }

    override fun build(): Person = Person(
        name = name,
        height = height,
        hairColor = hairColor,
        nationality = nationality
    )
}
